### PR TITLE
NAS-102992 / 11.3 / Fix ldap failover for SMB

### PR DESCRIPTION
**Pull request by anodos325**
- properly configure idmap settings and ldap.conf in presence of multiple LDAP URIs
- add support to get nslcd status from nslcd command socket
- fix DS cache format for LDAP and NIS

### CODE REVIEW

File: src/middlewared/middlewared/etc_files/local/nslcd.conf [hunk 1]
```diff
@@ -3,50 +3,50 @@
...
                 else:
                     certpath = cert['certificate_path']
         else:
             ldap = None
 
-        ldap_enabled = ldap['ldap_enable']
+        ldap_enabled = ldap['enable']
 
-        ldap_uri = "%s://%s" % ("ldaps" if ldap['ldap_ssl'] == "on" else "ldap", ldap['ldap_hostname'])
+        ldap_uri = ["%s://%s" % ("ldaps" if ldap['ssl'] == "on" else "ldap", x) for x in ldap['hostname']]
```
Comment (william-gr) [inline src/middlewared/middlewared/etc_files/local/nslcd.conf]: According to ldap.update, ssl should be ON, OFF or START_TLS

Is .config returning lowercase? if so we should fix that

File: src/middlewared/middlewared/plugins/ldap.py [hunk 2]
```diff
@@ -28,6 +34,87 @@ class SSL(enum.Enum):
     USETLS = 'START_TLS'
 
 
+class NlscdConst(enum.Enum):
+    NSLCD_CONF_PATH = '''/usr/local/etc/nslcd.conf'''
+    NSLCD_PIDFILE = '''/var/run/nslcd.pid'''
+    NSLCD_SOCKET = '''/var/run/nslcd/nslcd.ctl'''
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/ldap.py]: Why triple quote? thats odd : P

Comment (anodos325) [reply src/middlewared/middlewared/plugins/ldap.py]: It's in sample tools in nss-pam-ldapd. It bugs me too so I'll remove it.

File: src/middlewared/middlewared/etc_files/local/nslcd.conf (hunk 1 above)
Comment (anodos325) [reply src/middlewared/middlewared/etc_files/local/nslcd.conf]: Ah, that error crept in due to switching from datastore.query to ldap.config. Good catch.

File: src/middlewared/middlewared/etc_files/local/nss_ldap.conf [hunk 3]
```diff
@@ -6,38 +6,33 @@
...
+ssl ${ldap['ssl']}.lower()
         % if certpath:
 tls_cacert ${certpath}
         % endif
 tls_reqcert allow
     % endif
 scope sub
-timelimit 30
+timelimit ${ldap['timeout']}
 bind_timelimit 30
```
Comment (william-gr) [inline src/middlewared/middlewared/etc_files/local/nss_ldap.conf]: In nslcd.conf dns_timeout was used. Is this on purpose?

File: src/middlewared/middlewared/etc_files/local/openldap/ldap.conf [hunk 4]
```diff
@@ -12,40 +22,46 @@
...
 % if (ldap_enabled and ldap) or (ad_enabled and ad):
+# This file is used by Samba and py-ldap. If NETWORK_TIMEOUT is too high, then ldap failover
+# in Samba's ldapsam passdb backend may not occur.
 URI ${uri}
 BASE ${base}
+NETWORK_TIMEOUT ${network_timeout}
+TIMEOUT ${timeout}
     % if ssl:
-SSL ${ssl}
+# SSL ${ssl}
```
Comment (william-gr) [inline src/middlewared/middlewared/etc_files/local/openldap/ldap.conf]: Is this comment also on purpose?

File: src/middlewared/middlewared/plugins/ldap.py [hunk 5]
```diff
@@ -493,6 +580,19 @@ def get_dn(self, dn=None, ldap=None):
     async def __set_state(self, state):
         await self.middleware.call('cache.put', 'LDAP_State', state.name)
 
+    @private
+    def get_nslcd_status(self):
+        """
+        Returns internal nslcd state. nslcd will preferentially use the first LDAP server,
+        and only failover if the current LDAP server is unreachable.
+        """
+        ctx = NslcdClient(NlscdConst.NSLCD_ACTION_STATE_GET.value)
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/ldap.py]: Not important, but it seems a context manager for NslcdClient would be a perfect fit for this use case : P

Comment (anodos325) [reply src/middlewared/middlewared/plugins/ldap.py]: Good idea. I'll do that.

File: src/middlewared/middlewared/plugins/ldap.py [hunk 6]
```diff
@@ -28,6 +34,93 @@ class SSL(enum.Enum):
...
+        # write a request header with a request code
+        self.action = action
+        self.write_int32(NlscdConst.NSLCD_VERSION.value)
+        self.write_int32(action)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        self.close()
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/ldap.py]: Actually we should prob check for an exception and raising it after close

Comment (william-gr) [reply src/middlewared/middlewared/plugins/ldap.py]: Hm, I may be having a brainfart, I feel we talked about this before. Is the exception raised by default?

**Pull request merged**